### PR TITLE
feat(SnapDropZone): enable object stacking- fixes #1503

### LINKED ISSUE #1503: Proposal: Stackable objects of same type in the same SnapDropZone
## Overview

I'm proposing the addition of new functionality to `VRTK_SnapDropZone` to accommodate the stacking of identical `VRTK_InteractableObject` based items.

This will be useful for me in particular in building an RPG style inventory, but I can imagine it being useful in other contexts too (eg. tracking grenade count in an FPS chest-mounted inventory a la Onward)

## Code In Progress

- Made rough working changes to VRTK_SnapDropZone to tally the number of objects being stacked
- Added a (very) rough test to determine whether an interactable object is the same kind already snapped to an SDZ.
- Tweaked VRTK_InteractableObject to add a new "stackable" property

## Demo

capture

## For discussion

- Is this functionality generalized enough to merit inclusion into VRTK?
- What is the best way to perform interactable object type comparison (tags, subclass comparison, id variables?)
- Should all of the proposed functionality, like the item counter, be added to VRTK_SnapDropZone or should it be added to a subclassed prefab instead?
- What are the implications for communication between SDZs. Users might want to prevent multiple stacks of the same object, meaning that an SDZ would have to check in with each of the other SDZs in its group to see if they should receive the interactable object instead. Perhaps a VRTK_SnapDropZoneGroup component is worth considering.

### PR DESCRIPTION
**Pull request by bkensey**
SnapDropZones can now hold multiple InteractableObjects that share the same identifier. When in stack-mode SDZs now allow other objects to pass through currently held objects, which makes it easier to put additional objects into the SDZ without dealing with between-object collisions.

I haven't yet added a text component to the SDZ prefab. At present, an SDZ in stack mode will look to find a child Text object and make use of that, which is available for testing in example 41. 

This is an initial version and deserves some review and discussion before merging.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 1ffad02631ecea7384e7961ae40e0a1b63030a87
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by bkensey** [comment]
I have a few thoughts that I want to throw out there. I've been thinking a lot about how to make snap drop zones more flexible and extendable. I've been working on this PR with an eye towards eventually using SDZs in an RPG style inventory grid, with all of the item grouping, sorting, and snap re-routing that would entail.

At present, three different modes of snapping are emerging in my mind: 1) simple snapping 2) clone-on-unsnap and 3)stack snapping (which the rudimentary functionality already included in this PR)

Shoehorning all of this logic into the same functions is proving to make the conditional tests increasingly complex and therefore bug prone. What would everyone think about splitting the existing SDZ code into a more basic SDZ object that is responsible for handling object life cycle, highlight states, and event handling, with the snap logic delegated to a separate  "SnapAttachMechanic" that is defined as a property of every SDZ?

This would have the benefit of separating the code for "cloning-on-unsnap" and "stack-snapping" from each other and from simple snapping. The current SDZ code is just so interconnected and hard to parse out. I can't help but think more abstraction and separation would keep each use case more focused, and would keep SDZs from being cluttered with properties that are only relevant to one of the snap attach mechanics. A simple SDZ doesn't need the cloneOnUnsnap property, nor does a simple SDZ need all of the stack-snap properties I'm pondering.

The downside that making so many radical changes to SDZs will leave obsolete functions and properties all over the place. Would it be worth it? Thoughts?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 1ffad02631ecea7384e7961ae40e0a1b63030a87
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 1ffad02631ecea7384e7961ae40e0a1b63030a87
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

 * Commit: 3f2d1d6168299c63e9bfa5502f75716a509868bc
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 1ffad02631ecea7384e7961ae40e0a1b63030a87
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

 * Commit: 3f2d1d6168299c63e9bfa5502f75716a509868bc
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 7f4a0ee2984f9d5fa26d89fd1b599435839d3f54
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 1ffad02631ecea7384e7961ae40e0a1b63030a87
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

 * Commit: 3f2d1d6168299c63e9bfa5502f75716a509868bc
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 7f4a0ee2984f9d5fa26d89fd1b599435839d3f54
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by thestonefox** [comment]
this pull request is getting a bit messy. could probably do with a squash or something to clean up the commits into a concise chunk of work.

**Comment by bkensey** [comment]
Absolutely. I've been giving myself a crash course in interactive rebasing all night ;)

**Comment by bkensey** [comment]
Aaaaand I think everything is straight. Ready for review.

File: Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs [hunk 1]
```diff
@@ -63,10 +64,16 @@ public enum SnapTypes
         [Tooltip("The Snap Type to apply when a valid interactable object is dropped within the snap zone.")]
         public SnapTypes snapType = SnapTypes.UseKinematic;
         [Tooltip("The amount of time it takes for the object being snapped to move into the new snapped position, rotation and scale.")]
+        [HideInInspector]
```
Comment (thestonefox) [inline Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs]: Don't use `[HideInInspector]` for obsolete public attributes, use the new custom `[ObsoleteAttribute]` tag as this will still show the parameter in the inspector but change the label appearance to identify it as being deprecated.

Also delete the `[Tooltip()]` associated with the obsolete parameter, then the `[ObsoleteAttribute]` will use the `[Obsolete()]` message for the tooltip, which is more appropriate.

File: Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs [hunk 2]
```diff
@@ -104,25 +114,13 @@ public enum SnapTypes
         public event SnapDropZoneEventHandler ObjectUnsnappedFromDropZone;
 
         protected GameObject previousPrefab;
-        protected GameObject highlightContainer;
+        public GameObject highlightContainer { get; private set;}
```
Comment (thestonefox) [inline Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs]: If this requires a getter, then use a public method called `GetHighlightContainer()` instead of setting a protected method to public like this.

File: Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs [hunk 3]
```diff
@@ -154,7 +152,7 @@ public virtual void OnObjectSnappedToDropZone(SnapDropZoneEventArgs e)
 
         public virtual void OnObjectUnsnappedFromDropZone(SnapDropZoneEventArgs e)
         {
-            UnsnapObject();
+            snapAttachMechanicScript.UnsnapObject();
```
Comment (thestonefox) [inline Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs]: need to check that `snapAttachMechanicScript` is not null before performing any actions on it, in case it's been set to null at runtime, to prevent exception errors.

File: Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs [hunk 4]
```diff
@@ -186,65 +184,15 @@ public virtual void InitaliseHighlightObject(bool removeOldObject = false)
             GenerateObjects();
         }
 
-        /// <summary>
-        /// the ForceSnap method attempts to automatically attach a valid GameObject to the snap drop zone.
-        /// </summary>
-        /// <param name="objectToSnap">The GameObject to attempt to snap.</param>
-        public virtual void ForceSnap(GameObject objectToSnap)
```
Comment (thestonefox) [inline Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs]: A bunch of public methods have been removed without deprecating them. This is going to break a bunch of stuff for people. The public methods should be marked as Obsolete and the the internals of the method should be changed to do the new functionality.

File: Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs [hunk 5]
```diff
@@ -507,11 +338,11 @@ protected virtual void SetObjectHighlight()
             }
         }
 
-        protected virtual void ToggleHighlightColor()
+        public virtual void ToggleHighlightColor()
```
Comment (thestonefox) [inline Assets/VRTK/Prefabs/SnapDropZone/VRTK_SnapDropZone.cs]: public methods need to be grouped with other public methods above all protected methods, and public methods require documentation.

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_BaseSnapAttach.cs [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+
+            ForceUnsnap();
+            UnregisterAllGrabEvents();
+        }
+
+        /// <summary>
+        /// The IsSnapped method returns true if an Interactable Object is currently snapped to the Snap Drop Zone paired with this attach script
+        /// </summary>
+        /// <returns>Whetehr an Interactable Object is currently snapped to the Snap Drop Zone paired with this attach script.</returns>
+        public virtual bool IsSnapped()
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_BaseSnapAttach.cs]: public methods should be listed before protected.

goes public, protected, private

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_CloneSnapAttach.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
+﻿// Base Snap Attach|SnapAttachMechanics
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_CloneSnapAttach.cs]: This needs updating to what it actually is, also there is no ordering number at the end, so it will break.

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_BaseSnapAttach.cs [hunk 8]
```diff
@@ -0,0 +1,592 @@
+﻿// Base Snap Attach|SnapAttachMechanics
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_BaseSnapAttach.cs]: Needs an ordering number at the end

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_SimpleSnapAttach.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
+﻿// Base Snap Attach|SnapAttachMechanics
+namespace VRTK.SnapAttachMechanics
+{
+    using UnityEngine;
+
+    /// <summary>
+    /// Provides basic snapping functionality and serves as the default mechanic for a `VRTK_SnapDropZone'.
+    /// </summary>
+    /// <remarks>
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_SimpleSnapAttach.cs]: No remarks. Needs to include steps on how to use the script.

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_SimpleSnapAttach.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    [ExecuteInEditMode]
+    public class VRTK_SimpleSnapAttach : VRTK_BaseSnapAttach
+    {
+
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_SimpleSnapAttach.cs]: unnecessary line break

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_SimpleSnapAttach.cs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+                }
+            }
+           
+            //force reset isSnapped if the item is grabbed but isSnapped is still true
+            isSnapped = (isSnapped && interactableObjectCheck != null && interactableObjectCheck.IsGrabbed() ? false : isSnapped);
+            willSnap = !isSnapped;
+            wasSnapped = false;
+        }
+
+        public override void UnsnapObject()
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_SimpleSnapAttach.cs]: public before protected and requires documentation

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_StackSnapAttach.cs [hunk 12]
```diff
@@ -0,0 +1,249 @@
+﻿// Base Snap Attach|SnapAttachMechanics
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_StackSnapAttach.cs]: needs updating and ordering numbers

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_StackSnapAttach.cs [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+    using System.Collections;
+    using System.Collections.Generic;
+    using UnityEngine;
+    using UnityEngine.Assertions;
+    using UnityEngine.UI;
+
+    /// <summary>
+    /// Provides snapping and unsnapping behavior that permits stacking of identical Interactable Objects. Built on top of the basic Simple Snap Attach mechanic.
+    /// </summary>
+    /// <remarks>
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_StackSnapAttach.cs]: needs steps to use

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_StackSnapAttach.cs [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+        protected bool isResnapping = false;
+        protected Stack<GameObject> snapStack = new Stack<GameObject>();
+      
+        protected override void OnEnable()
+        {
+            base.OnEnable();
+            initializeSnapCountText();
+        }
+
+        public override void CheckCanSnap(VRTK_InteractableObject interactableObjectCheck)
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/SnapAttachMechanics/VRTK_StackSnapAttach.cs]: public before protected and docs required

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/VRTK_InteractableObject.cs [hunk 15]
```diff
@@ -198,6 +200,12 @@ public enum ValidDropTypes
         [Tooltip("Determines which controller can initiate a use action.")]
         public AllowedController allowedUseControllers = AllowedController.Both;
 
+        [Header("Stack Settings")]
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/VRTK_InteractableObject.cs]: Add line break after Header for consistency

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/VRTK_InteractableObject.cs [hunk 16]
```diff
@@ -198,6 +200,12 @@ public enum ValidDropTypes
...
+        [Tooltip("Determines if the object can be grouped with identical objects within a `VRTK_SnapDropZone`.")]
+        public bool isStackable = false;
+        [Tooltip("Used to determine whether objects are the same and therefore stackable within a `VRTK_SnapDropZone`.")]
+        public string objectID = "defaultInteractableObjectID";
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/VRTK_InteractableObject.cs]: wouldn't a policy list be better here rather than a rigid string?

Then you could determine things to be the same based on tag/layer/script applied.

Also, I don't like that this extra data is needed to be added to the Interactable Object. I feel it should be in a separate script so the Interactable Object doesnt get fatter than what it already is.

File: Assets/VRTK/Source/Scripts/Interactions/Interactables/VRTK_InteractableObject.cs [hunk 17]
```diff
@@ -716,6 +726,50 @@ public virtual void SaveCurrentState()
             }
         }
 
+        /// <summary>
+        /// The SaveColliderStates method stores the isTrigger states of the object's colliders.
+        /// </summary>
+        public virtual void SaveColliderStates(bool overwriteOriginalState)
```
Comment (thestonefox) [inline Assets/VRTK/Source/Scripts/Interactions/Interactables/VRTK_InteractableObject.cs]: Incomplete documentation, doesn't include the params

**Comment by bkensey** [comment]
I'll get to work on these. Thank you so much for the review!

**Comment by thestonefox** [comment]
I'm going to start closing these, we can reopen them if anyone raises it
